Xcode 15: Fix runtime warning for uploading media files

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11354 
<!-- Id number of the GitHub issue this PR addresses. -->

## What
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds a fix to get rid of the runtime warning for uploading media files on Xcode 15.

## Why
We were sending a non-empty `httpBody` for the upload requests. The compiler complains as it only processes the uploaded data, and the data sent in the `httpBody` is ignored.

## How
The fix is not to encode anything to send to the body of the upload requests. Specifically, I made the `encoding` optional for `DotcomRequest` to skip encoding request body for upload requests.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
Prerequisite: to confirm the fix, please build and run the app on Xcode 15.
- Log in to a store and navigate to the Products tab.
- Select any existing product or create a new one.
- Upload at least one photo from the device and save the product.
- The photo should be uploaded successfully to the store media library and the product should be saved successfully. No runtime warning should be displayed on Xcode.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->
N/A

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Make encoding optional for DotcomRequest

itsmeichigo committed:
Send nil encoding for upload requests

itsmeichigo committed:
Update release notes

itsmeichigo committed:
Remove redundant parameter context for uploadMedia

itsmeichigo committed:
Add tests for uploadMedia and uploadMediaToWordPressSite

selanthiraiyan reviewed (commented):
Thanks for working on this, Huong!

I tried testing this using application password login and see the following error while trying to upload image for a product.

```
The request of a upload task should not contain a body or a body stream, use `uploadTask(with:fromFile:)` or supply the body stream through the `urlSession(_:needNewBodyStreamForTask:)` delegate method.
```

itsmeichigo committed:
Add option to skip encoding request body for RESTRequest

itsmeichigo committed:
Add tests for skipsEncodingBody

itsmeichigo committed:
Merge branch 'trunk' into issue/11354-runtime-warning-upload-request

itsmeichigo commented:
Nice catch @selanthiraiyan! I added a new property to skip encoding body for RESTRequests in 6d4d9b45e15f763f504505fdf78a0ba7baee3ea1 and added new tests in a414d75f4769a77606ec7b3994aaf9b61e4a1af3. This PR is ready for another look.

selanthiraiyan reviewed (commented):
Thanks for fixing it, @itsmeichigo!

To discuss a different approach, I think making the `parameters` value `nil` in both `DotcomRequest` and `RESTRequest` will prevent the body from having unnecessary value. 

I tried the above solution in [this branch](https://github.com/woocommerce/woocommerce-ios/compare/try/parameter-nil?expand=1). Kindly let me know your thoughts.

itsmeichigo committed:
Revert "Add tests for skipsEncodingBody"

This reverts commit a414d75f4769a77606ec7b3994aaf9b61e4a1af3.

itsmeichigo committed:
Revert "Add option to skip encoding request body for RESTRequest"

This reverts commit 6d4d9b45e15f763f504505fdf78a0ba7baee3ea1.

itsmeichigo committed:
Remove nil encoding for upload requests

itsmeichigo committed:
Keep nil parameter as default for RESTRequest

itsmeichigo committed:
Keep encoding non-optional and skip encoding if parameters are nil

itsmeichigo committed:
Do not set empty parameters as default for DotcomRequest

itsmeichigo committed:
Update tests for RESTRequest

itsmeichigo committed:
Update tests for DotcomRequest

itsmeichigo committed:
Merge branch 'trunk' into issue/11354-runtime-warning-upload-request

itsmeichigo commented:
Thanks @selanthiraiyan! I've pushed new changes following your suggestions - please take a look at them when you can.

selanthiraiyan commented on Networking/Networking/Requests/RESTRequest.swift:
```diff
@@ -88,6 +88,9 @@ struct RESTRequest: Request {
         let request = try URLRequest(url: url, method: method)
+        guard let parameters else {
+            return request
+        }
```
I think we don't have to return here because the `encode` methods from `JSONEncoding` and  `URLEncoding` will return the `urlRequest` when `parameters` is nil.

itsmeichigo replied:
Good point! Updated in 90fcb74f9948848c27b2251fe81495b3dccb2fd3.

selanthiraiyan commented on Networking/Networking/Requests/DotcomRequest.swift:
```diff
@@ -96,7 +97,11 @@ struct DotcomRequest: Request, RESTRequestConvertible {
+            return try encoding.encode(dotcomRequest, with: parameters)
+        } else {
+            return dotcomRequest
+        }
```
Same as the other comment. I think we don't have to include this change as `encoding.encode(dotcomRequest, with: parameters)` won't alter the request if `parameters` is nil.

itsmeichigo replied:
Also updated in 90fcb74f9948848c27b2251fe81495b3dccb2fd3.

selanthiraiyan commented on Networking/Networking/Requests/DotcomRequest.swift:
```diff
@@ -69,6 +69,7 @@ struct DotcomRequest: Request, RESTRequestConvertible {
     ///     - path: RPC that should be executed.
     ///     - parameters: Collection of String parameters to be passed over to our target RPC.
     ///     - headers: Headers used in the URLRequest
+    ///     - encoding: How the parameters are encoded. Default to use `URLEncoding`.
```
We can remove this comment.

itsmeichigo replied:
I think we can keep this because `encoding` is a param in the initializer so it should be documented.

selanthiraiyan replied:
Oh right. 👍

selanthiraiyan reviewed (commented):
Thanks for the updates, Huong! 

I added a comment about always calling the `encode` method without having to check the `parameters`. I suggest this because `encode` method of different encoding types return the request without adding body when `parameters` is nil.

Kindly let me know your thoughts.

itsmeichigo committed:
Remove checks for nil parameters

itsmeichigo committed:
Merge branch 'trunk' into issue/11354-runtime-warning-upload-request

selanthiraiyan reviewed (approved):
Thanks for the updates, Huong! 🎉
